File: backend/app/routers/conversation.py
from typing import Any

from fastapi import APIRouter
from pydantic import BaseModel, Field

from app.services.conversation_store import (
    append_message,
    delete_conversation_note_document,
    get_conversation,
    get_conversation_note_task_ids,
    list_conversations,
    soft_delete_conversation,
    update_message,
    upsert_conversation,
)
from app.services.note_document_store import delete_note_task_artifacts
from app.services.note_task_store import cancel_note_task
from app.utils.response import ResponseWrapper as R

router = APIRouter()


class ConversationPayload(BaseModel):
    id: str
    mode: str = "chat"
    title: str | None = None
    status: str | None = None
    message: str | None = None
    platform: str | None = None
    linkedNoteTaskId: str | None = None
    noteState: str | None = None
    formData: dict[str, Any] = Field(default_factory=dict)
    transcript: dict[str, Any] = Field(default_factory=dict)
    audioMeta: dict[str, Any] = Field(default_factory=dict)
    markdown: Any = ""


class ConversationPatchPayload(BaseModel):
    mode: str | None = None
    title: str | None = None
    status: str | None = None
    message: str | None = None
    platform: str | None = None
    linkedNoteTaskId: str | None = None
    noteState: str | None = None
    formData: dict[str, Any] | None = None
    transcript: dict[str, Any] | None = None
    audioMeta: dict[str, Any] | None = None
    markdown: Any | None = None


class ConversationMessagePayload(BaseModel):
    id: str
    role: str
    message_type: str = "assistant_text"
    content: str
    status: str | None = None
    meta: dict[str, Any] = Field(default_factory=dict)
    createdAt: str | None = None
    updatedAt: str | None = None
    sources: list[dict[str, Any]] = Field(default_factory=list)
    error: bool = False


class ConversationMessagePatchPayload(BaseModel):
    role: str | None = None
    message_type: str | None = None
    content: str | None = None
    status: str | None = None
    meta: dict[str, Any] | None = None
    updatedAt: str | None = None
    sources: list[dict[str, Any]] | None = None
    error: bool | None = None


@router.get("/conversations")
def get_conversations():
    return R.success(list_conversations())


@router.get("/conversations/{conversation_id}")
def get_conversation_detail(conversation_id: str):
    item = get_conversation(conversation_id)
    if item is None:
        return R.error("会话不存在", code=404)
    return R.success(item)


@router.put("/conversations/{conversation_id}")
def put_conversation(conversation_id: str, data: ConversationPayload):
    payload = data.model_dump()
    payload["id"] = conversation_id
    return R.success(upsert_conversation(payload))


@router.patch("/conversations/{conversation_id}")
def patch_conversation(conversation_id: str, data: ConversationPatchPayload):
    current = get_conversation(conversation_id)
    if current is None:
        return R.error("会话不存在", code=404)

    payload = {
        "id": conversation_id,
        "mode": current.get("mode", "chat"),
        "title": current.get("title"),
        "status": current.get("status"),
        "message": current.get("message"),
        "platform": current.get("platform"),
        "linkedNoteTaskId": current.get("linkedNoteTaskId"),
        "noteState": current.get("noteState"),
        "formData": current.get("formData") or {},
        "transcript": current.get("transcript") or {},
        "audioMeta": current.get("audioMeta") or {},
        "markdown": current.get("markdown", ""),
    }
    updates = data.model_dump(exclude_unset=True)
    for field in ("formData", "transcript", "audioMeta"):
        if field in updates and isinstance(updates[field], dict):
            payload[field] = {
                **(payload.get(field) or {}),
                **updates.pop(field),
            }
    payload.update(updates)
    return R.success(upsert_conversation(payload))


@router.post("/conversations/{conversation_id}/messages")
def post_conversation_message(conversation_id: str, data: ConversationMessagePayload):
    try:
        return R.success(append_message(conversation_id, data.model_dump()))
    except ValueError as exc:
        message = str(exc)
        if message == "conversation not found":
            return R.error("会话不存在", code=404)
        return R.error(message, code=400)


@router.patch("/conversations/{conversation_id}/messages/{message_id}")
def patch_conversation_message(conversation_id: str, message_id: str, data: ConversationMessagePatchPayload):
    payload = data.model_dump(exclude_unset=True)
    payload["id"] = message_id
    try:
        return R.success(update_message(conversation_id, message_id, payload))
    except ValueError as exc:
        message = str(exc)
        if message == "message not found":
            return R.error("消息不存在", code=404)
        return R.error(message, code=400)


@router.delete("/conversations/{conversation_id}")
def delete_conversation(conversation_id: str):
    current = get_conversation(conversation_id)
    task_ids = get_conversation_note_task_ids(conversation_id)
    deleted = soft_delete_conversation(conversation_id)
    if not deleted:
        return R.error("会话不存在", code=404)
    for task_id in task_ids:
        delete_note_task_artifacts(task_id)
        cancel_note_task(task_id, "会话已删除，任务已取消")
    return R.success({"id": conversation_id})


@router.delete("/conversations/{conversation_id}/documents/{task_id}")
def delete_conversation_document(conversation_id: str, task_id: str):
    item = delete_conversation_note_document(conversation_id, task_id)
    if item is None:
        return R.error("文档不存在或会话不存在", code=404)
    return R.success(item)

File: frontend/src/services/conversation.ts
import request from '@/utils/request'

export interface ConversationPayload {
  id: string
  mode: 'chat' | 'note'
  title?: string
  status?: string
  message?: string
  platform?: string
  linkedNoteTaskId?: string
  noteState?: 'none' | 'generating' | 'ready' | 'failed'
  formData?: Record<string, any>
  transcript?: Record<string, any>
  audioMeta?: Record<string, any>
  markdown?: any
}

export type ConversationPatchPayload = Partial<Omit<ConversationPayload, 'id'>>

export interface ConversationDocumentPayload {
  taskId: string
  title: string
  content?: string
  sourceUrl?: string
  platform?: string
  modelName?: string
  style?: string
  status?: string
  wikiStatus?: string
  createdAt?: string
  updatedAt?: string
}

export interface ConversationMessagePayload {
  id: string
  role: 'user' | 'assistant' | 'system'
  message_type?: 'user_input' | 'assistant_text' | 'note_progress' | 'note_result' | 'system_error'
  content: string
  status?: 'pending' | 'running' | 'success' | 'failed'
  meta?: Record<string, any>
  createdAt?: string
  updatedAt?: string
  sources?: Array<Record<string, any>>
  error?: boolean
}

export interface ConversationImportPayload {
  import_mode: 'chat_asset' | 'note'
  conversation_id?: string
  title?: string
  content: string
  format?: string
  file_name?: string
  source_url?: string
  source_type?: string
  tags?: string[]
  metadata?: Record<string, any>
}

export interface ConversationImportResult {
  conversation_id: string
  conversation_title: string
  import_mode: 'chat_asset' | 'note'
  asset_id?: string
  note_id?: string
  document_task_id?: string
  message_id?: string
  status: string
  wiki_status: string
}

export const fetchConversations = async () => {
  return await request.get<any, ConversationPayload[]>('/conversations')
}

export const fetchConversation = async (conversationId: string) => {
  return await request.get<
    any,
    ConversationPayload & {
      messages?: ConversationMessagePayload[]
      documents?: ConversationDocumentPayload[]
      activeDocumentTaskId?: string
    }
  >(
    "/conversations/" + conversationId,
  )
}

export const upsertConversation = async (
  conversationId: string,
  payload: ConversationPayload,
) => {
  return await request.put<any, ConversationPayload>("/conversations/" + conversationId, payload)
}

export const patchConversation = async (
  conversationId: string,
  payload: ConversationPatchPayload,
) => {
  return await request.patch<any, ConversationPayload>("/conversations/" + conversationId, payload)
}

export const appendConversationMessage = async (
  conversationId: string,
  payload: ConversationMessagePayload,
) => {
  return await request.post<any, ConversationPayload & { messages?: ConversationMessagePayload[] }>(
    "/conversations/" + conversationId + "/messages",
    payload,
  )
}

export const patchConversationMessage = async (
  conversationId: string,
  messageId: string,
  payload: Partial<ConversationMessagePayload>,
) => {
  return await request.patch<any, ConversationPayload & { messages?: ConversationMessagePayload[] }>(
    "/conversations/" + conversationId + "/messages/" + messageId,
    payload,
  )
}

export const deleteConversation = async (conversationId: string) => {
  return await request.delete<any, { id: string }>("/conversations/" + conversationId)
}

export const deleteConversationDocument = async (conversationId: string, taskId: string) => {
  return await request.delete<
    any,
    ConversationPayload & {
      messages?: ConversationMessagePayload[]
      documents?: ConversationDocumentPayload[]
      activeDocumentTaskId?: string
    }
  >("/conversations/" + conversationId + "/documents/" + taskId)
}

export const importConversationMarkdown = async (payload: ConversationImportPayload) => {
  return await request.post<any, ConversationImportResult>('/notes/import', payload)
}
